app/org/sagebionetworks/bridge/services/ConsentServiceImpl.java

```diff
@@ -258,48 +269,38 @@ public void emailConsentAgreement(final Study study, final User user) {
+    private boolean hasUserSignedActiveConsent(String healthCode, SubpopulationGuid subpopGuid) {
+        checkNotNull(healthCode);
+        checkNotNull(subpopGuid);
 
-    @Override
-    public boolean isStudyAtEnrollmentLimit(Study study) {
-        if (study.getMaxNumOfParticipants() == 0) {
-            return false;
-        }
-        String key = RedisKey.NUM_OF_PARTICIPANTS.getRedisKey(study.getIdentifier());
-
-        long count = Long.MAX_VALUE;
-        String countString = jedisOps.get(key);
-        if (countString == null) {
-            // This is expensive but don't lock, it's better to do it twice slowly, than to throw an exception here.
-            count = userConsentDao.getNumberOfParticipants(study.getStudyIdentifier());
-            jedisOps.setex(key, TWENTY_FOUR_HOURS, Long.toString(count));
+        UserConsent userConsent = userConsentDao.getActiveUserConsent(healthCode, subpopGuid);
+        StudyConsentView mostRecentConsent = studyConsentService.getActiveConsent(subpopGuid);
+        
+        if (mostRecentConsent != null && userConsent != null) {
+            return userConsent.getConsentCreatedOn() == mostRecentConsent.getCreatedOn();
         } else {
-            count = Long.parseLong(countString);
-        }
-        return (count >= study.getMaxNumOfParticipants());
-    }
-
-    void incrementStudyEnrollment(Study study) throws StudyLimitExceededException {
-        if (study.getMaxNumOfParticipants() == 0) {
-            return;
-        }
-        if (isStudyAtEnrollmentLimit(study)) {
-            throw new StudyLimitExceededException(study);
+            return false;
         }
-        String key = RedisKey.NUM_OF_PARTICIPANTS.getRedisKey(study.getIdentifier());
-        jedisOps.incr(key);
     }
-
-    void decrementStudyEnrollment(Study study) {
-        if (study.getMaxNumOfParticipants() == 0) {
-            return;
-        }
-        String key = RedisKey.NUM_OF_PARTICIPANTS.getRedisKey(study.getIdentifier());
-        String count = jedisOps.get(key);
-        if (count != null && Long.parseLong(count) > 0) {
-            jedisOps.decr(key);
+    
+    private void updateSessionConsentStatuses(User user, SubpopulationGuid subpopGuid, boolean consented) {
+        if (user.getConsentStatuses() != null) {
+            List<ConsentStatus> updatedStatuses = Lists.newArrayList();
+            
+            for (ConsentStatus status : user.getConsentStatuses().values()) {
+                if (status.getSubpopulationGuid().equals(subpopGuid.getGuid())) {
+                    ConsentStatus updatedStatus = new ConsentStatus.Builder().withConsentStatus(status)
+                            .withConsented(consented).withSignedMostRecentConsent(consented).build();
+                    updatedStatuses.add(updatedStatus);
+                } else {
+                    updatedStatuses.add(status);
+                }
+            }
+            user.setConsentStatuses(ConsentStatus.toMap(updatedStatuses));
```

Comment:
I feel like this code would be much more elegant with a temporary list instead of a temporary map.